core/src/main/java/edu/wpi/grip/core/sources/VideoFileSource.java

```diff
@@ -0,0 +1,194 @@
+package edu.wpi.grip.core.sources;
+
+import edu.wpi.grip.core.Source;
+import edu.wpi.grip.core.events.SourceHasPendingUpdateEvent;
+import edu.wpi.grip.core.events.SourceRemovedEvent;
+import edu.wpi.grip.core.sockets.OutputSocket;
+import edu.wpi.grip.core.sockets.SocketHint;
+import edu.wpi.grip.core.sockets.SocketHints;
+import edu.wpi.grip.core.util.ExceptionWitness;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.eventbus.EventBus;
+import com.google.common.eventbus.Subscribe;
+import com.google.inject.assistedinject.Assisted;
+import com.google.inject.assistedinject.AssistedInject;
+import com.thoughtworks.xstream.annotations.XStreamAlias;
+
+import org.bytedeco.javacpp.opencv_core.Mat;
+import org.bytedeco.javacv.FFmpegFrameGrabber;
+import org.bytedeco.javacv.Frame;
+import org.bytedeco.javacv.FrameGrabber;
+import org.bytedeco.javacv.OpenCVFrameConverter;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A source for a video file input.
+ */
+@XStreamAlias("grip:VideoFile")
+public class VideoFileSource extends Source {
```

Comment:
There are some fundamental differences between them. Camera frame grabbers don't get null frame mats from the conversion; video files do. The grabber for the video file needs to update at a specific rate specified by the file itself; camera sources just grab as fast as possible.

And I'm using a Thread (scheduled executor now) to grab the frames because of these two issues.